Add velocity and story sprint to report

kalongthewires opened the pull request:
Not super familiar with the code base so this is very likely not the ideal way to do this. Just let me know. :)

Closes #24

khamper-nj committed:
WIP: Add velocity and story sprint to report

mewdriller commented on src/redux/selectors.js:
```diff
@@ -1,3 +1,23 @@
+
+const getVelocity = (stories, sprint) =>
+  stories.reduce((result, story) => {
+    if (isWithinRange(story.closedAt, sprint.startsAt, sprint.endsAt)) {
```
~~I've run into some problems through with needing to align the `date-fns` methods so that they don't consider time when evaluating comparisons. This _may_ be another situation where we need to create a `isWithinDayRange` variant that aligns the inputs to start-of-day.~~

~~There are several examples of that before (and a helper called `alignToDay`) in the `src/util.js` file.~~

☝️ this is resolved in fb62324a9fa61c2e97978147078ef35e8906d431 now.

mewdriller replied:
On a related note, I just double-check that it's inclusive of the start and end parameters; which I think is the behavior we want. So we're good on that front.

mewdriller replied:
☝️ this is resolved in fb62324a9fa61c2e97978147078ef35e8906d431 now.

mewdriller commented on src/redux/selectors.js:
```diff
@@ -1,3 +1,23 @@
+    isWithinRange(story.closedAt, sprint.startsAt, sprint.endsAt),
+  );
+
+  return storySprint ? storySprint.number : 'None';
```
🎨: Maybe `get(find(...), 'number', 'None')` here since we've got `lodash` around?

mewdriller commented on src/redux/selectors.js:
```diff
@@ -1,3 +1,23 @@
+      return result + story.estimate;
+    }
+
+    return result;
```
🎨: Maybe an implicit return arrow in the `reduce` for `isWithinRange() ? result + story.estimate : result`?

mewdriller commented on src/redux/selectors.js:
```diff
@@ -1,3 +1,23 @@
+import { isWithinRange } from 'date-fns';
+import { find } from 'lodash';
+
+const getVelocity = (stories, sprint) =>
```
From looking at the b2io/scopebuilder + AEP/TLE report, it appears there may be an issue with calculating velocity for the first sprint when there were stories closed before it started. The 0th tick on the graph shows 198 closed, the 1st tick shows 221 closed, but the report lists 17 as the velocity for Sprint 1 rather than 23. 

![image](https://user-images.githubusercontent.com/492553/35922130-c2c3e216-0bea-11e8-8f85-1784b526762d.png)

mewdriller replied:
Looking further in, there's issues that closed on the day the sprint ended that aren't counted as being in that sprint.

![image](https://user-images.githubusercontent.com/492553/35922779-72cad3d0-0bec-11e8-80d6-ed32fd50940a.png)

![image](https://user-images.githubusercontent.com/492553/35922873-b96dc216-0bec-11e8-8cb9-440fb2c668fc.png)

That six points would account for the 17 to 23 offset. This probably is a start-of-day alignment issue looking at it.

mewdriller replied:
⚠️: This also makes me wonder about an edge-case where a story is closed _between_ sprints when the sprints don't include the weekends (like in b2io/scopebuilder).

mewdriller commented:
~~Pretty sure this is unrelated but just noticed an issue with a suspicious close date. We probably didn't close "scss vendors import order" on 1969-12-31. That happens to be the start-of-day aligned unix epoch date so we probably have something nil being treated as a `0` and getting converted to a date then start-of-day aligned.~~

☝️this is resolved in fb62324a9fa61c2e97978147078ef35e8906d431 now.

![image](https://user-images.githubusercontent.com/492553/35922443-9ce2fa86-0beb-11e8-99eb-18e78624acba.png)

mewdriller reviewed (changes requested):
To figure out what to suggest you do I had to dig in a little bit. I'll push a small commit here in a sec that resolves a couple of the issues.

There's still the ⚠️ edge-case about stories closing between sprints and a couple of 🎨comments. 

Overall it LGTM. Let me know what you think. 😄 

:high_brightness: :hibiscus:

mewdriller committed:
WIP: Resolve date alignment issues

- Catch null dates in shortDay util
- Align date range checks in selector to start-of-day

mewdriller commented:
Just realized I hit "Request Changes" somehow rather than "Approve". Sorry about that!

mewdriller committed:
Merge branch 'master' into 24-add-velocity-to-report
